vassal-app/src/main/java/VASSAL/build/GameModule.java

```diff
@@ -829,21 +864,97 @@ public JToolBar getToolBar() {
   }
 
   /**
-   * Append the string to the title of the controls window and all Map windows
-   * @param s If null, set the title to the default.
+   * Returns an appropriate Title Bar string for a window, based on the module name,
+   * the last read/written game file, and the manner of interaction with it.
+   * @return appropriate title bar string for a window.
+   * @param key Localization key to be used to generate string
+   * @param name Name of the object whose title bar is to be generated
    */
-  public void appendToTitle(String s) {
-    if (s == null) {
-      frame.setTitle(Resources.getString("GameModule.frame_title", getLocalizedGameName()));  //$NON-NLS-1$
+  public String getWindowTitleString (String key, String name) {
+    if (StringUtils.isEmpty(gameFile) || GameFileMode.NEW_GAME.equals(gameFileMode)) {
+      return Resources.getString(key + "_title", name);  //NON-NLS-1$
+    }
+    else if (GameFileMode.SAVED_GAME.equals(gameFileMode)) {
+      return Resources.getString(key + "_title_saved", name, gameFile); //NON-NLS-1$
+    }
+    else if (GameFileMode.LOADED_GAME.equals(gameFileMode)) {
+      return Resources.getString(key + "_title_loaded", name, gameFile); //NON-NLS-1$
+    }
+    else if (GameFileMode.REPLAYED_GAME.equals(gameFileMode)) {
+      return Resources.getString(key + "_title_replayed", name, gameFile); //NON-NLS-1$
+    }
+    else if (GameFileMode.REPLAYING_GAME.equals(gameFileMode)) {
+      return Resources.getString(key + "_title_replaying", name, gameFile); //NON-NLS-1$
+    }
+    else if (GameFileMode.LOGGING_GAME.equals(gameFileMode)) {
+      return Resources.getString(key + "_title_logging", name, gameFile); //NON-NLS-1$
+    }
+    else if (GameFileMode.LOGGED_GAME.equals(gameFileMode)) {
+      return Resources.getString(key + "_title_logged", name, gameFile); //NON-NLS-1$
     }
     else {
-      frame.setTitle(frame.getTitle() + s);
+      return "**Unknown Game File Mode** - " + gameFile;
     }
```

Comment:
You could replace the entire contents of this function with this, by taking advantage of the string names of the enum values:
```
    if (StringUtils.isEmpty(gameFile) || GameFileMode.NEW_GAME.equals(gameFileMode)) {
      return Resources.getString(key + "_title", name);  //NON-NLS-1$
    }
    else {
      return Resources.getString(key + "_title_" + gameFileMode, name, gameFile); //NON-NLS-1$
    }
```

Reply:
We don't want the extremal case to be possible, right? The only way that can happen is if somehow `gameFileMode` has gotten set to null. I think we should just disallow that---throw an NPE in `setGameFile` and `setGameFileMode` if someone tries to set it null.